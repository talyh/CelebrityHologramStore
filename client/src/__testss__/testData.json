[
    {
        "_id": "5b53fd760f3f76191b4d548b",
        "pictureURL": "https://m.media-amazon.com/images/M/MV5BMTI1MjU3MTI2MF5BMl5BanBnXkFtZTcwMDgxNTE4MQ@@._V1_UY317_CR1,0,214,317_AL_.jpg",
        "detailsURL": "https://www.imdb.com/name/nm0000095/?ref_=fn_al_nm_1",
        "name": "Woody Allen",
        "roles": [
            "Writer",
            "Director",
            "Actor"
        ]
    },
    {
        "_id": "5b54017a0f3f76191b4d5490",
        "name": "Ryan Reynolds",
        "pictureURL": "https://m.media-amazon.com/images/M/MV5BOTI3ODk1MTMyNV5BMl5BanBnXkFtZTcwNDEyNTE2Mg@@._V1_UY317_CR6,0,214,317_AL_.jpg",
        "detailsURL": "https://www.imdb.com/name/nm0005351/?ref_=nv_sr_1",
        "roles": [
            "Actor"
        ]
    },
    {
        "_id": "5b565cef0489c37162da83cb",
        "name": "Scarlett Johansson",
        "roles": [
            "Actress"
        ],
        "pictureURL": "https://m.media-amazon.com/images/M/MV5BMTM3OTUwMDYwNl5BMl5BanBnXkFtZTcwNTUyNzc3Nw@@._V1_UY317_CR23,0,214,317_AL_.jpg",
        "detailsURL": "https://www.imdb.com/name/nm0424060/?ref_=nv_sr_1"
    },
    {
        "_id": "5b565d280489c37162da83cd",
        "name": "Natalie Portman",
        "roles": [
            "Actress"
        ],
        "pictureURL": "https://m.media-amazon.com/images/M/MV5BMTQ3ODE3Mjg1NV5BMl5BanBnXkFtZTcwNzA4ODcxNA@@._V1_UY317_CR11,0,214,317_AL_.jpg",
        "detailsURL": "https://www.imdb.com/name/nm0000204/?ref_=nv_sr_1"
    },
    {
        "_id": "5b565dbf0489c37162da83ce",
        "name": "Neil Patrick Harris",
        "roles": [
            "Actor",
            "Producer"
        ],
        "pictureURL": "https://m.media-amazon.com/images/M/MV5BMTUxNjQ4MjU5Ml5BMl5BanBnXkFtZTcwMTY5NzAzNw@@._V1_UY317_CR14,0,214,317_AL_.jpg",
        "detailsURL": "https://www.imdb.com/name/nm0000439/?ref_=nv_sr_1"
    },
    {
        "_id": "5b566cba0489c37162da83d5",
        "name": "Hugh Laurie",
        "roles": [
            "Actor"
        ],
        "pictureURL": "https://m.media-amazon.com/images/M/MV5BNDYzNjAzMTYwOF5BMl5BanBnXkFtZTcwMTc3NTcxOA@@._V1_UY317_CR17,0,214,317_AL_.jpg",
        "detailsURL": "https://www.imdb.com/name/nm0491402/?ref_=nv_sr_3"
    },
    {
        "_id": "5b566f010489c37162da83d6",
        "name": "Steven Spielberg",
        "roles": [
            "Director"
        ],
        "pictureURL": "https://m.media-amazon.com/images/M/MV5BMTY1NjAzNzE1MV5BMl5BanBnXkFtZTYwNTk0ODc0._V1_UX214_CR0,0,214,317_AL_.jpg",
        "detailsURL": "https://www.imdb.com/name/nm0000229/?ref_=nv_sr_1"
    },
    {
        "_id": "5b5670380489c37162da83d7",
        "name": "Tom Hanks",
        "roles": [
            "Actor"
        ],
        "pictureURL": "https://m.media-amazon.com/images/M/MV5BMTQ2MjMwNDA3Nl5BMl5BanBnXkFtZTcwMTA2NDY3NQ@@._V1_UY317_CR2,0,214,317_AL_.jpg",
        "detailsURL": "https://www.imdb.com/name/nm0000158/?ref_=nv_sr_2"
    },
    {
        "_id": "5b5675970489c37162da83dc",
        "name": "Robin Wright",
        "roles": [
            "Actress"
        ],
        "pictureURL": "https://m.media-amazon.com/images/M/MV5BMTU0NTc4MzEyOV5BMl5BanBnXkFtZTcwODY0ODkzMQ@@._V1_UY317_CR4,0,214,317_AL_.jpg",
        "detailsURL": "https://www.imdb.com/name/nm0000705/?ref_=nv_sr_4"
    },
    {
        "_id": "5b5675d70489c37162da83dd",
        "name": "Tom Cruise",
        "roles": [
            "Actor"
        ],
        "pictureURL": "https://m.media-amazon.com/images/M/MV5BMTk1MjM3NTU5M15BMl5BanBnXkFtZTcwMTMyMjAyMg@@._V1_UY317_CR14,0,214,317_AL_.jpg",
        "detailsURL": "https://www.imdb.com/name/nm0000129/?ref_=nv_sr_1"
    }
]