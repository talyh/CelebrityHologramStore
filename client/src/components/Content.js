import styled from "styled-components"

// provide a styled wrapper for the whatever is to be displayed in the App
const Content = styled.div`
  position: relative;
  top: 12em;
  padding: 2em;
`

export default Content