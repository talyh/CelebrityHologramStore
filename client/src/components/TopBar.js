import React from "react"
import Header from "./innerPieces/Header"
import AppTitle from "./innerPieces/AppTitle"
import AppSubtitle from "./innerPieces/AppSubTitle"
import SearchBar from "./SearchBar"

const TopBar = ({ title, subtitle, onSearch, onTitleClick }) => {
    return (
        <Header id="header">
            <AppTitle id="title" onClick={onTitleClick}>{title}</AppTitle>
            <AppSubtitle id="subtitleArea">{subtitle}</AppSubtitle>
            <SearchBar onSearch={onSearch} />
        </Header>
    )
}

export default TopBar