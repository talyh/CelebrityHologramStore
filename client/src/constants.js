export const pageModes = Object.freeze({ "list": 1, "details": 2, "insertion": 3 })

export const cardModes = Object.freeze({ "small": 1, "big": 2 })