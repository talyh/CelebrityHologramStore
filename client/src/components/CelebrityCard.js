import React from "react"
import styled from "styled-components"
import Icon from "./generic/Icon"
import { cardModes } from "../constants"

export const StyledCard = styled.div`
    display: grid;
    grid-template-rows: 2fr 1fr;
    background-color: #E8E8E8;
    -webkit-box-shadow: 8px 8px 6px -6px #777777;
    -moz-box-shadow: 8px 8px 6px -6px #777777;
     box-shadow: 8px 8px 6px -6px #777777;
    cursor:${props => props.onClick ? "pointer" : "normal"};
    height:${props => props.cellHeight}px;
    color: #707070;

    &:hover {
         transform: scale(${props => props.hoverScale}); 
      }
`

const StyledCelebrityCard = StyledCard.extend`
    display: grid;
    grid-template-rows: 1fr 1fr 1fr;
    grid-template-columns: 2fr 2fr 2fr;
    grid-template-areas: "Picture Name Name"
                        "Picture Roles Roles"
                        "Picture . Remove";
    column-gap: 1em;
    padding: 1.5em;
`

const StyledName = styled.h1`
    grid-area: Name;
    font-size: ${props => props.mode === cardModes.small ? "1em" : "2em"};
    pointer-events: none;
`

const StyledRoles = styled.div`
    grid-area: Roles;
    font-size: ${props => props.mode === cardModes.small ? "1em" : "2em"};
    color: #B9B9B9;
    pointer-events: none;
`

const StyledImage = styled.img`
    grid-area: Picture;
    width: ${props => props.mode === cardModes.small ? "100px" : "300px"}
    pointer-events: none;
`

const RemoveIcon = styled(Icon).attrs({
    id: "removeIcon",
    src: "remove.png",
    alt: "Remove",
    style: {
        paddingTop: "1.3em",
        paddingBottom: "0.7em",
        marginLeft: "0.5em",
        marginRight: "0.5em",
        "gridArea": "Remove"
    }
})`
`

const CelebrityCard = ({ celebrity, callbackForRemove, mode, onHover, onClick, hoverScale, cellHeight, rolesInPreview }) => {
    const remove = criteria => fetch(`http://localhost:3001/celebrities?id=${criteria}`, { method: 'DELETE' })
        .then(response => response.ok ? response.json() : handleError(response))
        .then(callbackForRemove)
        .catch(error => handleError(error))

    const handleError = error => console.log(error)

    const condenseList = (itemsToShow, list) => {
        let listSize = itemsToShow

        if (itemsToShow > list.length) {
            listSize = list.length
        }

        let condensedList = []

        for (let i = 0; i < listSize; i++) {
            const newItem = <div id={list[i]} key={list[i]}>{list[i]}</div>
            condensedList = [...condensedList, newItem]
        }

        if (rolesInPreview < list.length) {
            const showMore = <div id="More roles" key="moreRoles"> {`${list.length - itemsToShow} more`}</ div>
            condensedList = [...condensedList, showMore]
        }

        return condensedList
    }


    const showRoles = () => {
        const roles = [...celebrity.roles].sort()
        if (mode === cardModes.small) {
            return <StyledRoles mode={mode} id="roles">{condenseList(rolesInPreview, roles)}</StyledRoles>
        }
        else {
            return <StyledRoles mode={mode} id="roles">{condenseList(roles.length, roles)}</StyledRoles>
        }
    }

    return (
        <StyledCelebrityCard id={celebrity._id} onClick={onClick} onMouseEnter={onHover && (ev => onHover(ev.target.id))} hoverScale={hoverScale} cellHeight={cellHeight} >
            <StyledName id={celebrity.name} mode={mode}>{celebrity.name}</StyledName>
            {showRoles()}
            <StyledImage alt={`${celebrity.name}`} src={celebrity.pictureURL} mode={mode} />
            {mode === cardModes.big && <div>Find out more @ <a href={celebrity.detailsURL}>IMDB</a></div>}
            <RemoveIcon onClick={() => remove(celebrity._id)} />
        </StyledCelebrityCard >
    )
}

export default CelebrityCard