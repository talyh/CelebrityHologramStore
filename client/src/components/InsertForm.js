import React, { Component } from "react"
import styled from "styled-components"
import Card from "./innerPieces/CelebrityCardContainer"
import TextInput from "./generic/TextInput"
import Picture from "./innerPieces/CelebrityPicture"
import { AddIcon, RemoveIcon } from "./innerPieces/Icons"
import { SaveButton, CancelButton } from "./innerPieces/Buttons"
import { cardModes } from "../constants";

const PictureInputArea = styled.div`
    grid-Area: Picture;
`

const NameInputArea = styled.div`
    grid-Area: Name;
`

const RolesInputArea = styled.div`
    grid-Area: Roles;
    margin-bottom: 1em;
`

const DetailsInputArea = styled.div`
    grid-Area: More;
`

const ButtonsArea = styled.div`
    grid-Area: Buttons;
    display: grid;
    grid-template-rows: 1fr;
    grid-template-columns: repeat(2, 1fr);
    column-gap: 1em;
`

class InsertForm extends Component {
    state = {
        name: "",
        picture: {
            url: "",
            validURL: false
        },
        detailsURL: "",
        roles: [""]
    }

    changeName = ev => this.setState({ name: ev.target.value })
    changePicture = ev => this.setState({ picture: { validURL: this.validatePictureURL(ev.target.value), url: ev.target.value } })
    changeDetails = ev => this.setState({ detailsURL: ev.target.value })
    addRole = () => this.setState(prevState => ({ roles: [...prevState.roles, ""] }))
    changeRole = (index, ev) => this.setState({ roles: [...this.state.roles.slice(0, index), ev.target.value, ...this.state.roles.slice(index + 1)] })
    removeRole = (index, ev) => this.setState({ roles: [...this.state.roles.slice(0, index), ...this.state.roles.slice(index + 1)] })
    validatePictureURL = url => {
        const img = new Image()
        let result = false
        img.onerror = img.onabort = () => result = false
        img.onload = () => result = true
        img.src = url
        return result
    }

    save = () => fetch(`http://localhost:3001/celebrities?name=${this.state.name}&roles=${JSON.stringify(this.state.roles)}&pictureURL=${this.state.picture.url}&detailsURL=${this.state.detailsURL}`, { method: 'POST' })
        .then(response => response.ok ? response.json() : this.handleError(response))
        .then(this.props.onSave)
        .catch(error => this.handleError(error))

    handleError = error => console.log(error)

    render() {
        return (
            <Card mode={cardModes.insertion}>
                <NameInputArea>
                    <TextInput placeholder="Enter a celebrity name" value={this.state.name} onChange={this.changeName} />
                </NameInputArea>
                <PictureInputArea>
                    <TextInput placeholder="Enter a picture URL" value={this.state.picture.url} onChange={this.changePicture} />
                    <Picture
                        src={this.state.picture.url ? this.state.picture.url : "/imgs/blank.png"}
                        alt={this.state.picture.validURL ? "Image selected" : "No image selected"} />
                </PictureInputArea>
                <RolesInputArea>
                    <AddIcon small alt="Add Role" onClick={this.addRole} id="roles" style={{ paddingRight: "0.5em", marginBottom: "0.5em", verticalAlign: "middle" }} />
                    Click to add more roles
                    {this.state.roles.map((role, index) => <div key={index}>
                        <TextInput placeholder="Enter a role" value={role} onChange={ev => this.changeRole(index, ev)} width="70%" />
                        {index > 0 && <RemoveIcon alt="Remove Role" onClick={ev => this.removeRole(index, ev)} />}
                    </div>)}
                </RolesInputArea>
                <DetailsInputArea>
                    <TextInput placeholder="Enter an IMDB URL" value={this.state.detailsURL} onChange={this.changeDetails} />
                </DetailsInputArea>
                <ButtonsArea>
                    <CancelButton onClick={this.props.onCancel} />
                    <SaveButton onClick={this.save} />
                </ButtonsArea>
            </Card>
        )
    }
}

export default InsertForm