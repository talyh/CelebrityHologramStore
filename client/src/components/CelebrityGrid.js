import React from "react"
import CelebrityCard from "./CelebrityCard"

const CelebrityGrid = ({ celebrityList, add }) => {

    const dummyStyle = { cursor: "pointer" }

    const generateCards = array => array.map(entry => <CelebrityCard celebrity={entry} key={entry._id} />)

    return <div>
        <div style={dummyStyle} onClick={add}>Add</div>
        {generateCards(celebrityList)}
    </div>
}

export default CelebrityGrid