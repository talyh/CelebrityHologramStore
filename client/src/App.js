import React, { Component } from "react"
import styled from "styled-components"
import Header from "./components/Header"
import CelebrityGrid from "./components/CelebrityGrid"
import CelebrityCard from "./components/CelebrityCard"
import InsertForm from "./components/InsertForm"
import { pageModes, cardModes } from "./constants"

const Content = styled.div`
  position: relative;
  top: 12em;
  padding: 2em;
`

class App extends Component {
  state = {
    mode: "",
    celebrityList: [],
    celebrityShown: ""
  }

  setCelebrityList = newList => this.setState({ celebrityList: newList })

  refreshCelebrityList = () => {
    fetch("http://localhost:3001/celebrities")
      .then(response => response.ok ? response.json() : this.handleError(response))
      .then(result => this.setCelebrityList(result))
      .catch(error => this.handleError(error))
  }

  showList = () => { this.setState({ mode: pageModes.list }); this.refreshCelebrityList() }
  showInsertForm = () => this.setState({ mode: pageModes.insertion })
  showDetails = () => this.setState({ mode: pageModes.details })
  selectCelebrity = id => this.setState({ celebrityShown: this.state.celebrityList.filter(celebrity => celebrity._id === id)[0] })

  componentDidMount() {
    this.showList()
  }

  handleError = error => console.log(error)

  render() {
    const list = <CelebrityGrid
      celebrityList={this.state.celebrityList}
      add={this.showInsertForm}
      remove={this.refreshCelebrityList}
      onCardHover={this.selectCelebrity}
      onCardClick={this.showDetails} />
    const details = <CelebrityCard
      celebrity={this.state.celebrityShown}
      callbackForRemove={this.showList}
      mode={cardModes.big} />
    const insertForm = <InsertForm
      onCancel={this.showList}
      onSave={this.showList} />

    const determineContents = () => {
      switch (this.state.mode) {
        case pageModes.list:
        default:
          return list
        case pageModes.details:
          return details
        case pageModes.insertion:
          return insertForm
      }
    }

    return (
      <div>
        <Header
          title="Celebrity Hologram Store"
          subtitle="Your favorite celebrities in a single place"
          onSearch={this.setCelebrityList}
          onTitleClick={this.showList} />
        <Content id="content">
          {determineContents()}
        </Content>
      </div>
    )
  }
}

export default App
