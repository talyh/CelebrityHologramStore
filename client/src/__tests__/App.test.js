import React from 'react'
import ReactDOM from 'react-dom'
import App from '../App'
import { shallow } from "enzyme"
import { pageModes, cardModes } from "../constants"
import TopBar from "../components/TopBar"
import Content from "../components/Content"
import CelebrityGrid from "../components/CelebrityGrid"
import CelebrityCard from "../components/CelebrityCard"
import InsertForm from "../components/InsertForm"
import testData from "./testData"
import get from "../operations/getRecords"

jest.mock("../operations/getRecords")

describe("Unit Tests", () => {
  describe("Rendering", () => {
    it('renders without crashing', () => {
      const div = document.createElement('div');
      ReactDOM.render(<App />, div);
      ReactDOM.unmountComponentAtNode(div);
    })

    it("matches snapshot", () => {
      const rendered = shallow(<App />)

      expect(rendered).toMatchSnapshot()
    })

    it("always has a top bar", () => {
      const rendered = shallow(<App />)

      const topBar = rendered.find(TopBar)

      expect(topBar).toHaveLength(1)
    })

    it("always has a content container", () => {
      const rendered = shallow(<App />)

      const content = rendered.find(Content)

      expect(content).toHaveLength(1)
    })

    it("content has a single child", () => {
      const rendered = shallow(<App />)

      const content = rendered.find(Content)

      expect(content.children).toHaveLength(1)
    })

    it("list mode shows a celebrity grid", () => {
      const rendered = shallow(<App />)
      rendered.setState({
        mode: pageModes.list,
      })

      const celebrityGrid = rendered.find(CelebrityGrid)

      expect(celebrityGrid).toHaveLength(1)
    })

    it("details mode shows a celebrity card", () => {
      const rendered = shallow(<App />)
      rendered.setState({
        mode: pageModes.details,
      })

      const celebrityCard = rendered.find(CelebrityCard)

      expect(celebrityCard).toHaveLength(1)
    })

    it("insertion mode shows an insertion form", () => {
      const rendered = shallow(<App />)
      rendered.setState({
        mode: pageModes.insertion,
      })

      const insertForm = rendered.find(InsertForm)

      expect(insertForm).toHaveLength(1)
    })
  })

  describe("Props passed down", () => {
    describe("TopBar", () => {
      it("setCelebrityList is passed down to the TopBar as onSearch", () => {
        const rendered = shallow(<App />)

        const topBar = rendered.find(TopBar)

        expect(topBar.prop("onSearch")).toBe(rendered.instance().setCelebrityList)
      })

      it("showList is passed down to the TopBar as onTitleClick", () => {
        const rendered = shallow(<App />)

        const topBar = rendered.find(TopBar)

        expect(topBar.prop("onTitleClick")).toBe(rendered.instance().showList)
      })
    })

    describe("CelebrityGrid", () => {
      it("celebrityList is passed down to the CelebrityGrid as celebrityList", () => {
        const rendered = shallow(<App />)
        rendered.setState({
          mode: pageModes.list,
          celebrityList: testData
        })

        const celebrityGrid = rendered.find(CelebrityGrid)

        expect(celebrityGrid.prop("celebrityList")).toBe(testData)
      })

      it("showInsertForm is passed down to the CelebrityGrid as add", () => {
        const rendered = shallow(<App />)
        rendered.setState({
          mode: pageModes.list
        })

        const celebrityGrid = rendered.find(CelebrityGrid)

        expect(celebrityGrid.prop("add")).toBe(rendered.instance().showInsertForm)
      })

      it("showList is passed down to the CelebrityGrid as close", () => {
        const rendered = shallow(<App />)
        rendered.setState({
          mode: pageModes.list
        })

        const celebrityGrid = rendered.find(CelebrityGrid)

        expect(celebrityGrid.prop("close")).toBe(rendered.instance().showList)
      })

      it("selectCelebrity is passed down to the CelebrityGrid as onCardHover", () => {
        const rendered = shallow(<App />)
        rendered.setState({
          mode: pageModes.list
        })

        const celebrityGrid = rendered.find(CelebrityGrid)

        expect(celebrityGrid.prop("onCardHover")).toBe(rendered.instance().selectCelebrity)
      })

      it("showDetails is passed down to the CelebrityGrid as onCardClick", () => {
        const rendered = shallow(<App />)
        rendered.setState({
          mode: pageModes.list
        })

        const celebrityGrid = rendered.find(CelebrityGrid)

        expect(celebrityGrid.prop("onCardClick")).toBe(rendered.instance().showDetails)
      })
    })

    describe("CelebrityCard", () => {
      it("celebrity object is passed down to the CelebrityCard as celebrity", () => {
        const rendered = shallow(<App />)
        rendered.setState({
          mode: pageModes.details,
          celebrityList: testData,
          celebritySelected: "5b53fd760f3f76191b4d548b"
        })

        const returned = rendered.instance().getCelebrity(rendered.state().celebritySelected)
        const celebrityCard = rendered.find(CelebrityCard)

        expect(celebrityCard.prop("celebrity")).toBe(returned)
      })

      it("showList is passed down to the CelebrityCard as callbackForClose", () => {
        const rendered = shallow(<App />)
        rendered.setState({
          mode: pageModes.details
        })

        const celebrityCard = rendered.find(CelebrityCard)

        expect(celebrityCard.prop("callbackForClose")).toBe(rendered.instance().showList)
      })

      it("showList is passed down to the CelebrityCard as callbackForRemove", () => {
        const rendered = shallow(<App />)
        rendered.setState({
          mode: pageModes.details
        })

        const celebrityCard = rendered.find(CelebrityCard)

        expect(celebrityCard.prop("callbackForRemove")).toBe(rendered.instance().showList)
      })

      it("cardMode.details is passed down to the CelebrityCard as mode", () => {
        const rendered = shallow(<App />)
        rendered.setState({
          mode: pageModes.details
        })

        const celebrityCard = rendered.find(CelebrityCard)

        expect(celebrityCard.prop("mode")).toBe(cardModes.details)
      })
    })

    describe("InsertForm", () => {
      it("showList is passed down to the InsertForm as onCancel", () => {
        const rendered = shallow(<App />)
        rendered.setState({
          mode: pageModes.insertion
        })

        const insertForm = rendered.find(InsertForm)

        expect(insertForm.prop("onCancel")).toBe(rendered.instance().showList)
      })

      it("showList is passed down to the InsertForm as onSave", () => {
        const rendered = shallow(<App />)
        rendered.setState({
          mode: pageModes.insertion
        })

        const insertForm = rendered.find(InsertForm)

        expect(insertForm.prop("onSave")).toBe(rendered.instance().showList)
      })
    })
  })

  describe("Operations", () => {
    describe("setCelebrityList", () => {
      it("setCelebrityList updates state.celebrityList with received array", () => {
        const rendered = shallow(<App />)

        expect(rendered.state().celebrityList).toEqual([])

        rendered.instance().setCelebrityList(testData)

        expect(rendered.state().celebrityList).toEqual(testData)
      })

      it("setCelebrityList doesn't update state if argument is not an array", () => {
        const rendered = shallow(<App />)

        expect(rendered.state().celebrityList).toEqual([])

        const argument = "argument"
        rendered.instance().setCelebrityList(argument)

        expect(rendered.state().celebrityList).toEqual([])
      })
    })

    describe("refreshCelebrityList", () => {
      it("refreshCelebrityList calls get with the proper arguments", () => {
        const rendered = shallow(<App />)

        rendered.instance().refreshCelebrityList()

        expect(get.mockImplementation.call.length).toBe(1)
        expect(get.mock.calls[0][0]).toBe("/celebrities")
        expect(get.mock.calls[0][1]).toEqual(expect.any(Function))
        expect(get.mock.calls[0][2]).toEqual(expect.any(Function))
      })
    })

    describe("showList", () => {
      it("showList updates state.mode with list", () => {
        const rendered = shallow(<App />)
        rendered.setState({
          mode: pageModes.details
        })

        rendered.instance().showList()

        expect(rendered.state().mode).toEqual(pageModes.list)
      })

      it("showList calls refreshList", () => {
        const rendered = shallow(<App />)
        const refresh = jest.spyOn(rendered.instance(), "refreshCelebrityList")

        rendered.instance().showList()

        expect(refresh).toBeCalled()
      })
    })

    describe("showInsertForm", () => {
      it("showInsertForm updates state.mode with insertion", () => {
        const rendered = shallow(<App />)

        rendered.instance().showInsertForm()

        expect(rendered.state().mode).toEqual(pageModes.insertion)
      })
    })

    describe("showDetails", () => {
      it("showDetails updates state.mode with details when celebrity is selected", () => {
        const rendered = shallow(<App />)
        rendered.setState({
          celebritySelected: "123"
        })

        rendered.instance().showDetails()

        expect(rendered.state().mode).toEqual(pageModes.details)
      })

      it("showDetails doesn't change state when celebrity is not selected", () => {
        const rendered = shallow(<App />)
        const originalState = rendered.state()

        rendered.instance().showDetails()

        expect(rendered.state()).toEqual(originalState)
      })
    })

    describe("selectCelebrity", () => {
      it("selectCelebrity updates state.celebritySelected with argument", () => {
        const rendered = shallow(<App />)
        const id = "123"

        rendered.instance().selectCelebrity(id)

        expect(rendered.state().celebritySelected).toEqual(id)
      })
    })

    describe("getCelebrity", () => {
      it("getCelebrity filters celebrityList based on argument", () => {
        const rendered = shallow(<App />)
        rendered.setState({
          celebrityList: [{ _id: "123" }, { _id: "456" }]
        })
        const id = "123"

        const returned = rendered.instance().getCelebrity(id)

        expect(returned).toEqual({ _id: "123" })
      })
    })

    describe("componentDidMount", () => {
      it("componentDidMount calls refreshList", () => {
        const rendered = shallow(<App />)
        const refresh = jest.spyOn(rendered.instance(), "refreshCelebrityList")

        rendered.instance().componentDidMount()

        expect(refresh).toBeCalled()
      })
    })
  })
})

describe("Functional Automated Test", () => {
  it("when successful, refreshCelebrityList updates the list with current entries", () => {
    const rendered = shallow(<App />)
    rendered.setState({
      celebrityList: []
    })

    rendered.instance().refreshCelebrityList()

    expect(rendered.state().celebrityList).toBe(testData)
  })
})