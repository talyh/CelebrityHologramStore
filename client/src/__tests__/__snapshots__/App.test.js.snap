// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <TopBar
          onSearch={[Function]}
          onTitleClick={[Function]}
          subtitle="Your favorite celebrities in a single place"
          title="Celebrity Hologram Store"
/>,
        <styled.div
          id="content"
>
          <CelebrityGrid
                    add={[Function]}
                    celebrityList={Array []}
                    close={[Function]}
                    onCardClick={[Function]}
                    onCardHover={[Function]}
                    remove={[Function]}
          />
</styled.div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "onSearch": [Function],
          "onTitleClick": [Function],
          "subtitle": "Your favorite celebrities in a single place",
          "title": "Celebrity Hologram Store",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <CelebrityGrid
            add={[Function]}
            celebrityList={Array []}
            close={[Function]}
            onCardClick={[Function]}
            onCardHover={[Function]}
            remove={[Function]}
/>,
          "id": "content",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "add": [Function],
            "celebrityList": Array [],
            "close": [Function],
            "onCardClick": [Function],
            "onCardHover": [Function],
            "remove": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <TopBar
            onSearch={[Function]}
            onTitleClick={[Function]}
            subtitle="Your favorite celebrities in a single place"
            title="Celebrity Hologram Store"
/>,
          <styled.div
            id="content"
>
            <CelebrityGrid
                        add={[Function]}
                        celebrityList={Array []}
                        close={[Function]}
                        onCardClick={[Function]}
                        onCardHover={[Function]}
                        remove={[Function]}
            />
</styled.div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "onSearch": [Function],
            "onTitleClick": [Function],
            "subtitle": "Your favorite celebrities in a single place",
            "title": "Celebrity Hologram Store",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <CelebrityGrid
              add={[Function]}
              celebrityList={Array []}
              close={[Function]}
              onCardClick={[Function]}
              onCardHover={[Function]}
              remove={[Function]}
/>,
            "id": "content",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "add": [Function],
              "celebrityList": Array [],
              "close": [Function],
              "onCardClick": [Function],
              "onCardHover": [Function],
              "remove": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
